proxy/src/outbound.rs

```diff
@@ -38,33 +41,52 @@ impl<B> Outbound<B> {
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum Destination {
+    LocalSvc(FullyQualifiedAuthority),
+    External(SocketAddr),
+}
+
 impl<B> Recognize for Outbound<B>
 where
     B: tower_h2::Body + 'static,
 {
     type Request = http::Request<B>;
     type Response = bind::HttpResponse;
     type Error = <Self::Service as tower::Service>::Error;
-    type Key = (FullyQualifiedAuthority, Protocol);
+    type Key = (Destination, Protocol);
     type RouteError = ();
     type Service = Buffer<Balance<
         load::WithPendingRequests<Discovery<B>>,
         choose::PowerOfTwoChoices<rand::ThreadRng>,
     >>;
 
     fn recognize(&self, req: &Self::Request) -> Option<Self::Key> {
-        req.uri().authority_part().map(|authority| {
-            let auth = FullyQualifiedAuthority::new(
+        let local = req.uri().authority_part().and_then(|authority| {
+            FullyQualifiedAuthority::new(
                 authority,
                 self.default_namespace.as_ref().map(|s| s.as_ref()),
-                self.default_zone.as_ref().map(|s| s.as_ref()));
-
-            let proto = match req.version() {
-                http::Version::HTTP_2 => Protocol::Http2,
-                _ => Protocol::Http1,
-            };
-            (auth, proto)
-        })
+                self.default_zone.as_ref().map(|s| s.as_ref()))
+
+        });
+
+        let dest = if let Some(local) = local {
+            Destination::LocalSvc(local)
+        } else {
+            let orig_dst = req.extensions()
+                .get::<Arc<ctx::transport::Server>>()
+                .and_then(|ctx| {
+                    ctx.orig_dst_if_not_local()
+                });
+            Destination::External(orig_dst?)
```

Comment:
It is wrong to fail here. If we can't avoid failing here in this PR then let's add a TODO here acknowledging the problem. (Presumably this problem will be fixed when we implement DNS lookups.)

Reply:
I don't see why it is wrong to fail here. If we have nothing that will allow us to route the request (recognize, as this method is called), then there's nothing to be done but to return an error.

This is also what is done in `Inbound`.

Reply:
It's wrong to fail here because, without the proxy, the request wouldn't have failed. In practice, we expect this kind of failure to never occur because we expect the proxy to only run on Linux and we expect `SO_ORIGINAL_DST` to work. Also, in the future we'll be able to make it work by doing a DNS lookup to get the address.